aequilibrae/project/network/link.py

```diff
@@ -1,25 +1,194 @@
+from aequilibrae.project.database_connection import database_connection
+from aequilibrae.project.network.mode import Mode
+from aequilibrae import logger
 
 
-class Link:
-    def __init__(self, link_id=None):
-        p = Parameters()
-        fields = p.parameters["network"]["links"]["fields"]
-        one_way_fields = [list(x.keys())[0] for x in fields["one-way"]]
-        one_way_fields = {x: None for x in one_way_fields}
+class Link(SafeClass):
+    """A Link object represents a single record in the *links* table
 
-        twf = [list(x.keys())[0] for x in fields["two-way"]]
-        two_way_fields = {f"{x}_ab": None for x in twf}
-        two_way_fields.update({f"{x}_ba": None for x in twf})
+    ::
 
-        one_way_fields.update(two_way_fields)
-        self.__dict__.update(one_way_fields)
+        from aequilibrae import Project
 
-        if link_id is not None:
-            self.link_id = link_id
-            self._populate()
+        proj = Project()
+        proj.open('path/to/project/folder')
 
-    def _populate(self):
-        print(self.link_id)
+        all_links = proj.network.links
+
+        # Let's get a mode to work with
+        modes = proj.network.modes
+        car_mode = modes.get('c')
+
+        # We can just get one link in specific
+        link1 = all_links.get(4523)
+        link2 = all_links.get(3254)
+
+        # We can find out which fields exist for the links
+        which_fields_do_we_have = link1.data_fields()
+
+        # And edit each one like this
+        link1.lanes_ab = 3
+        link1.lanes_ba = 2
+
+        # we can drop a mode from the link
+        link1.drop_mode(car_mode)
+        # or link1.drop_mode('c')
+
+        # we can add a mode to the link
+        link2.add_mode(car_mode)
+        # or link2.add_mode('c')
+
+        # Or set all modes at once
+        link2.set_modes('cmtw')
+
+        # We can just save the link
+        link1.save()
+        link2.save()
+        """
+
+    def __init__(self, dataset):
+        super().__init__(dataset)
+        self.__fields = list(dataset.keys())
+
+        self.__new = dataset['geometry'] is None
+
+    def delete(self):
+        """Deletes link from database"""
+        conn = database_connection()
+        curr = conn.cursor()
+        curr.execute(f'DELETE FROM links where link_id="{self.link_id}"')
+        conn.commit()
+        del self
+
+    def save(self):
+        """Saves link to database"""
+        conn = database_connection()
+        curr = conn.cursor()
+
+        data = []
+        if self.__new:
+            up_keys = []
+            for key, val in self.__dict__.items():
+                if key not in self.__original__ or key == 'geometry':
+                    continue
+                up_keys.append(f'"{key}"')
+                data.append(val)
+            markers = ','.join(['?'] * len(up_keys)) + ',GeomFromWKB(?)'
+            up_keys.append('geometry')
+            data.append(self.geometry.wkb)
```

Comment:
I think you know what i'm going to say about 30+ line if statements

```
if __new
   sql, data = get_sql_for_new_object()
else:
   sql,data = get_sql_for_update_object()
```

Reply:
Like this?

Reply:
yup - just like that